plugins/connection/redirection_utility.c

```diff
@@ -0,0 +1,555 @@
+/************************************************************************************
+	Copyright (C) 2015 MariaDB Corporation AB,
+
+   This library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Library General Public
+   License as published by the Free Software Foundation; either
+   version 2 of the License, or (at your option) any later version.
+
+   This library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Library General Public License for more details.
+
+   You should have received a copy of the GNU Library General Public
+   License along with this library; if not see <http://www.gnu.org/licenses>
+   or write to the Free Software Foundation, Inc.,
+   51 Franklin St., Fifth Floor, Boston, MA 02110, USA
+*************************************************************************************/
+
+#include <ma_global.h>
+#include <ma_sys.h>
+#include <mysql.h>
+#include <errmsg.h>
+#include <mysql/client_plugin.h>
+#include <string.h>
+#include <ma_common.h>
+#include <mariadb_ctype.h>
+#include "ma_server_error.h"
+#include <ma_pthread.h>
+#include "redirection_utility.h"
+
+// Note that key, host and user and malloced together with Redirection_Entry
+struct Redirection_Entry
```

Comment:
why don't you use  LIST (ma_list.c) ? For an example check mysql->stmts and ma_invalidate_stmts() function.

Reply:
I don't really think LIST is a good approximation here... Redirection cache is an LRU cache indexed by a string-value key, so I expect key addressing like C++ dictionaries which LIST justifiably does not support. Besides I need to move a list node to head or tail of that list(so that the least recent used record can be found immediately), which LIST does not support either.
I tried writing the cache as a LIST of key-indexed Redirection Entries and got a getter function like below:

	struct Redirection_Entry *cur_entry;
	struct Redirection_Entry *ret_entry;
    /* find the entry by key */
	LIST *cur = redirection_cache_head;
	while (cur)
	{
		cur_entry = (struct Redirection_Entry *)(cur->data);
		if (cur_entry && cur_entry->key)
		{
			if (!strcmp(cur_entry->key, key))
			{
				ret_entry = cur_entry;
				break;
			}
			cur = cur->next;
		}
		else break;
	}
	if (ret_entry)
	{
		/* move to list head */
		if (redirection_cache_head != cur)
		{
			if (redirection_cache_tail == cur){
				redirection_cache_tail = cur->prev;
			}
			cur->prev->next = cur->next;

			if (cur->next)
			{
				cur->next->prev = cur->prev;
			}

			cur->next = redirection_cache_head;
			cur->prev = 0;

			redirection_cache_head = cur;
		}
     }   
It occurs to me I didn't gain much from the LIST abstraction other than the prev and next pointer.

Reply:
On second thought, it might be a good idea to implement redirection cache as an array instead of a list. Array design is not ideal in that finding the least recent used cache entry takes O(n) time instead of O(1), but compared to time spent on network requests the overhead should be negligible. On the other hand, with an array of records there's no more complicated manipulation of head, tail, next and prev pointers, and (hopefully) my code will be much easier to read. May I have your opinion on which path should I take?

Reply:
Hi 9EOR9, I've re-implemented redirection cache as an array. Would you mind taking a look at the latest changes?